### PR TITLE
fix is_writable not working for NFS mounts

### LINKED ISSUE #40810: Php's `is_writable` function always returns `false` on NFS mounts
I opened the issue owncloud-docker/server#394 and tracked down the problem to be in the following piece of code:

https://github.com/owncloud/core/blob/12737e7a107101acffe9a4073dee9936b6566c3e/lib/private/legacy/util.php#L786-L797

The `is_writable` function seems to function improperly when operating on NFS mounts. Although the mount is writable by the executing user, Php falsely states that it is not writable. This has also been discussed in [this StackOverflow thread](https://stackoverflow.com/questions/50707811/php-is-writable-false-for-nfs-folder-although-files-can-be-written).

The problem already got address in NextCloud via nextcloud/server#13237. I hereby propose to merge those changes into the ownCloud core.

### PR DESCRIPTION
**Pull request by axelrindle**
## Description
For context see #40810

## Related Issue
- Fixes #40810

## Motivation and Context
I want to use ownCloud with it's data directory being an NFS mount, which currently does not work.

## How Has This Been Tested?
- "Production" setup on a Raspberry Pi 4 with Raspberry Pi OS
- Modified the `util.php` file locally
- Tested a fresh installation as well with upgrading an existing installation

## Screenshots (if appropriate):
None.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

File: lib/private/legacy/util.php [hunk 1]
```diff
@@ -784,16 +784,24 @@ public static function checkServer(\OCP\IConfig $config) {
...
-					['<a href="' . $urlGenerator->linkToDocs('admin-dir_permissions') . '" target="_blank" rel="noreferrer">', '</a>']
-				);
-				$errors[] = [
-					'error' => 'Your Data directory is not writable by ownCloud',
-					'hint' => $permissionsHint
-				];
+				// is_writable doesn't work for NFS mounts, so try to write a file and check if it exists
+				// see https://github.com/nextcloud/server/pull/13237
+				$testFile = sprintf('%s/%s.tmp', $CONFIG_DATADIRECTORY, uniqid('data_dir_writability_test_'));
+				$handle = @fopen($testFile, 'w');
```
Comment (phil-davis) [inline lib/private/legacy/util.php:790]: Note: I had to suppress `fopen` crashing out here - if it "goes wrong" we want to continue and cehck the return value (which is `false` if the file could not be opened)

File: tests/lib/UtilCheckServerTest.php [hunk 2]
```diff
@@ -147,19 +147,41 @@ public function testDataDirNotWritable() {
 			$this->markTestSkipped('You are running tests as root - this test will not work in this case.');
 		}
 
-		\chmod($this->datadir, 0300);
+		// Set read+execute access to the data directory
+		\chmod($this->datadir, 0500);
```
Comment (phil-davis) [inline tests/lib/UtilCheckServerTest.php:151]: 4 is read
2 is write
1 is execute

So "300" was write+execute for the owner (no read). That was a weird thing, and not what the unit test name suggests.
"500" is read+execute. On a folder, execute gives the ability to "read' the file names... in the folder.

File: tests/lib/UtilCheckServerTest.php [hunk 3]
```diff
@@ -147,19 +147,41 @@ public function testDataDirNotWritable() {
...
+	 */
+	public function testDataDirReadOnly() {
+		if ($this->getCurrentUser() === 'root') {
+			$this->markTestSkipped('You are running tests as root - this test will not work in this case.');
+		}
+
+		// set read-only access to the data directory (not even execute access)
+		// In this case, the process cannot even list the data dir contents
+		// So it cannot see the file '.ocdata' that is in the directory
+		\chmod($this->datadir, 0400);
```
Comment (phil-davis) [inline tests/lib/UtilCheckServerTest.php:171]: I added a test case for if the folder really only has read access (and no execute).
In this case, `checkServer` returns `$result` with 2 things that it complains about, which is fine.

File: tests/lib/UtilCheckServerTest.php [hunk 4]
```diff
@@ -147,19 +147,41 @@ public function testDataDirNotWritable() {
...
+		\chmod($this->datadir, 0700); //needed for cleanup
+		$this->assertCount(2, $result);
+	}
+
 	/**
 	 * Tests no error is given when the datadir is not writable during setup
 	 */
 	public function testDataDirNotWritableSetup() {
-		\chmod($this->datadir, 0300);
+		\chmod($this->datadir, 0500);
```
Comment (phil-davis) [inline tests/lib/UtilCheckServerTest.php:184]: I also fixed this test to have the folder access set to read+execute.

**Comment by phil-davis** [comment]
This looks OK to me.
@DeepDiver1975 @jvillafanez or anybody else, please review.

**Comment by axelrindle** [comment]
> I'm not an NFS expert, but is it possible that the issue comes from a wrong / incompatible NFS setup?

I don't think that this is the case. Access is performed exclusively by the `www-data` user. The webserver as well as the php-fpm process run as that user. I also did all of the manual testing with the `www-data` user.

Manually working in the NFS mapped directory using standard unix commands works perfectly fine.

**Comment by jvillafanez** [comment]
Could you check if the setup below works for you? There are things to adjust and maybe it isn't the most secure environment, but it's worthy to check.

NFS server , "/etc/exports" file
```
/opt *(rw,sync,fsid=0,all_squash,anonuid=1001,anongid=1001,no_subtree_check)
/opt/deletest *(rw,sync,all_squash,anonuid=1001,anongid=1001,no_subtree_check)
```
client
```
> sudo mount -t nfs4 -vvv <ip-addr>:/deletest /mnt
```
test file
```
<?php

$filename = '/mnt/README.md';

echo "$filename\n";
var_dump(is_writeable($filename));

file_put_contents($filename, "huehue" . date('c') ."\n");

$filename = '/mnt';

echo "$filename\n";
var_dump(is_writeable($filename));
```

Test file works for me, showing both the "/mnt" and "/mnt/README.md" paths as writable. If I didn't miss anything, ownCloud should be able to access to the NFS via the "/mnt" directory. I can't fully test it because I don't know why I'm having issues to mount the NFS inside a docker container, so I think I'll have to setup a full VM with ownCloud in it and that will take a lot of time.

**Comment by axelrindle** [comment]
Unfortunately, no. I'm running a Synology NAS where I cannot manually edit the `/etc/exports` file. In fact, I don't know if the Synology OS even has that file.

**Comment by jvillafanez** [comment]
It should be possible through the synology UI
https://kb.synology.com/en-us/DSM/tutorial/How_to_access_files_on_Synology_NAS_within_the_local_network_NFS
linking to https://kb.synology.com/en-us/DSM/help/DSM/AdminCenter/file_share_privilege_nfs?version=7

I think the important part is the "all_squash" option, which should match the "Map all users to guest" in synology

**Comment by axelrindle** [comment]
Choosing `Map all users to guest` seems to fix the issue. I span up a testing instance and will try the same for the production instance tomorrow. If that works too, there should be an addition to the documentation.

**Comment by jvillafanez** [comment]
Taking into account that ownCloud requires exclusive access to the data directory, the "Map all users to guest" option seems reasonable.
I don't know how NFS behaves if there is no mapping for the user, but maybe the problem was that the ownCloud's "www-data" user was mapped to an unexpected user in the NFS. If you're using multiple ownCloud servers in a cluster setup, maybe each of those ownCloud servers were accessing to the NFS with different users.

**Comment by axelrindle** [comment]
I did some extended testing and my production setup seems to work again. In the Synology interface I chose the option `Map root user to guest` because `Map all users to guest` resulted in no access at all because the `guest` user on the Synology system does not have any access to the owncloud folder. I'm afraid I can't exactly tell what the issue is, but I think it is rather to the NFS protocol itself and filesystem permissions.

Furthermore the `is_writable` function still seems to be buggy, I found the following bug ticket: [Bug #68926 | is_writable returns false but file_put_contents works](https://bugs.php.net/bug.php?id=68926).

If anyone is interested, I did the following to "fix" the issue:

- unmount the NFS directory on the client
- reset permissions on the NFS server
- as squash options choose `Map root user to guest`
- add the following to `/etc/fstab` (or edit existing) on the client:
```
192.168.178.5:/volume1/ownCloud     /home/www-data/mnt/owncloud nfs     rw,hard,noexec,nosuid    0   0
```
- run `mount -a` as `root` on the client
- run `chmod 770 /home/www-data/mnt/owncloud` as `root` on the client
- run `chown -R www-data:www-data /home/www-data/mnt/owncloud` as `root` on the client
- `php7.4 occ` does not spit out any errors anymore

I will close this for now.

**Pull request closed without merging**